drivers: mcux_flexspi: Removing Logging Statements from FlexSPI driver

danieldegrasse opened the pull request:
Program flow will behave in undefined ways if the ARM core in NXP  i.MXRT parts attempts to access the flash memory using XIP while the flexspi driver is writing to or reading from flash. The undefined behavior manifests in a variety of ways, such as in #39340, where a load from flexspi memory returns 0x0 (and the subsequent branch causes a usage fault), or in #40133, where the program loads a garbage memory address and branches to it.

The flexspi driver itself is relocated to ITCM at runtime using `zephyr_code_relocate` macros, but the driver makes calls to the logging subsystem, which is linked into XIP flash memory. The simplest solution I can see is to remove all logging statement from the flexspi driver. Alternatively, the logging subsystem (and any logging backends) could all be linked to ITCM, but this seems likely to end up introducing new errors, as any new logging subsystem would also need to be properly relocated to ITCM.

Fixes #40133
Fixes #39340

Laczen commented:
@danieldegrasse, would it not be simpler to disable logging in case of XIP:
```
#include <logging/log.h>
#ifdef CONFIG_FLASH_MCUX_FLEXSPI_XIP
LOG_MODULE_REGISTER(flexspi_hyperflash, 0);
#else
LOG_MODULE_REGISTER(flexspi_hyperflash, CONFIG_FLASH_LOG_LEVEL);
#endif
```

danieldegrasse commented:
@Laczen It would be simpler, thanks for the suggestion. I've switched the PR to use that approach.

MaureenHelm reviewed (changes requested):
Quietly ignoring `CONFIG_FLASH_LOG_LEVEL` could cause confusion for someone debugging this driver, it would be better to override the default instead. Something like this in `soc/arm/nxp_imx/rt/Kconfig.defconfig.series`:
```
# Avoid RWW hazards
config FLASH_LOG_LEVEL
	default 0 if FLASH_MCUX_FLEXSPI_XIP
	depends on FLASH
```

MaureenHelm commented:
And as an added safeguard, you may also want to add a build warning or error to the driver, in the event that someone sets `CONFIG_FLASH_LOG_LEVEL` in their prj.conf or at the command line:

```c
#if defined(FLASH_MCUX_FLEXSPI_XIP) && (CONFIG_FLASH_LOG_LEVEL > 0)
#warning "Enabling flash driver logging and XIP mode simultaneously can cause read-while-write hazards. This configuration is not recommended.
#endif
```

de-nordic commented:
> 
> 
> And as an added safeguard, you may also want to add a build warning or error to the driver, in the event that someone sets `CONFIG_FLASH_LOG_LEVEL` in their prj.conf or at the command line:
> 
> ```c
> #if defined(FLASH_MCUX_FLEXSPI_XIP) && (CONFIG_FLASH_LOG_LEVEL > 0)
> #warning "Enabling flash driver logging and XIP mode simultaneously can cause read-while-write hazards. This configuration is not recommended.
> #endif
> ```

Maybe we should consider this an error that can be squashed by special Kconfig option, something like "FLASH_LOG_I_KNOW_FLEXSPI_MAY_BREAK_WITH_IT", so that user would have to acknowledge that this might be problem.

danieldegrasse committed:
drivers: mcux_flexspi: Default logging to disabled when XIP is used

Program flow will behave incorrectly (memory and instruction fetches
return invalid data) if Flexspi is accessed by the Flexspi driver while
being used as XIP memory by the Cortex M7.

Set logging to disabled by when XIP mode is used in the memc and
flexspi drivers, and warn the user if they attempt to enable it.

Fixes #40133

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

bbilas commented on drivers/flash/flash_mcux_flexspi_hyperflash.c:
```diff
@@ -11,12 +11,25 @@
 #include <fsl_cache.h>
 #endif
 
+
```
It doesn't seem to be needed.

MaureenHelm reviewed (changes requested):
Code changes look good, but the commit message doesn't match anymore. Can you update it to say that you're disabling logging under certain conditions, rather than removing logging altogether?

danieldegrasse commented:
> 
> 
> Code changes look good, but the commit message doesn't match anymore. Can you update it to say that you're disabling logging under certain conditions, rather than removing logging altogether?

Fixed, sorry about that! Should have checked the commit message closer

MaureenHelm reviewed (approved):
>> Code changes look good, but the commit message doesn't match anymore. Can you update it to say that you're disabling logging under certain conditions, rather than removing logging altogether?
>
> Fixed, sorry about that! Should have checked the commit message closer

No problem. Thank you!
